File: src/main/java/com/cam/rxtest/Given.java
package com.cam.rxtest;

import rx.Observable;
import rx.Scheduler;
import rx.functions.Func1;
import rx.functions.Func2;

import java.util.Map;

public class Given<T, U> {

    private final ExecutionContext<T, U> context;

    public Given(ExecutionContext<T, U> context) {
        this.context = context;
    }

    public Given<T, U> theStreamUnderTest(Func1<Observable<T>, Observable<U>> f) {
        Observable<U> sut = f.call(context.sourceAsObservable());
        context.setStreamUnderTest(sut);
        return this;
    }

    public Given<T, U> theStreamUnderTest2(Func1<Map<String, ? extends Observable<T>>, Observable<U>> f) {
        Observable<U> sut = f.call(context.allSourcesAsObservables());
        context.setStreamUnderTest(sut);
        return this;
    }

    public Given<T, U> theStreamUnderTest(Func2<Observable<T>, Scheduler, Observable<U>> f) {
        Observable<U> sut = f.call(context.sourceAsObservable(), context.getScheduler());
        context.setStreamUnderTest(sut);
        return this;
    }

    public When<T, U> when() {
        return new When<>(context);
    }

    public Given<T, U> aSourceStream(String id) {
        context.source(id);
        return new Given<>(context);
    }
}

File: rxscala-test-dsl/src/main/scala/uk/camsw/rxscala/test/dsl/Given.scala
package uk.camsw.rxscala.test.dsl

import java.time

import rx.lang.scala
import rx.lang.scala.JavaConversions._
import rx.lang.scala.{subjects, Observable}
import rx.subjects.{Subject, PublishSubject}
import uk.camsw.rxjava.test.dsl.given.BaseGiven
import uk.camsw.rxjava.test.dsl.scenario.ExecutionContext

class Given[T1, U](ctx: ExecutionContext[T1, T1, U, Given[T1, U], When[T1, U]]) extends BaseGiven[U, Given[T1, U], When[T1, U]](ctx) {
  val context = ctx

  override def when(): When[T1, U] = context.getWhen

  def theStreamUnderTest(f: (Observable[T1], scala.Scheduler) => Observable[U]) = {
    val source: Observable[T1] = context.getSource1.asObservable()
    val sut: Observable[U] = f(source, context.getScheduler)
    context.setStreamUnderTest(toJavaObservable[U](sut))
    this
  }

  def theStreamUnderTest(f: () => Observable[U]) = {
    context.setStreamUnderTest(toJavaObservable[U](f()))
    this
  }

  def theCustomSource(customSource: PublishSubject[T1] ) = {
    context.setCustomSource1(customSource)
    this
  }
  

}